livekit-plugins/livekit-plugins-lmnt/livekit/plugins/lmnt/tts.py

```diff
@@ -0,0 +1,285 @@
+                        break
+
+            except asyncio.CancelledError:
+                if ws:
+                    await ws.close()
+                    if listen_task:
+                        await asyncio.shield(listen_task)
+                break
+            except Exception as e:
+                if retry_count > max_retry and max_retry > 0:
+                    logging.error(f"failed to connect to LMNT: {e}")
+                    break
+
+                retry_delay = min(retry_count * 5, 5)  # max 5s
+                retry_count += 1
+                logging.warning(
+                    f"failed to connect to LMNT: {e} - retrying in {retry_delay}s"
+                )
+                await asyncio.sleep(retry_delay)
+
+        self._closed = True
+
+    async def _try_connect(self) -> aiohttp.ClientWebSocketResponse:
+        ws = await self._session.ws_connect(f"{self._config.base_url}/ai/speech/stream")
+
+        init_packet = {
+            AUTHORIZATION_HEADER: self._config.api_key,
+            "voice": self._config.voice.id,
+        }
+        await ws.send_str(json.dumps(init_packet))
+        return ws
+
+    async def _listen_task(self, ws: aiohttp.ClientWebSocketResponse) -> None:
+        while True:
+            msg = await ws.receive()
+
+            if msg.type in (
+                aiohttp.WSMsgType.CLOSED,
+                aiohttp.WSMsgType.CLOSE,
+                aiohttp.WSMsgType.CLOSING,
+            ):
+                break
+
+            if msg.type != aiohttp.WSMsgType.BINARY:
+                continue
+
+            audio_frame = decode_mp3_to_frame(msg.data)
+            self._event_queue.put_nowait(
+                tts.SynthesisEvent(
+                    type=tts.SynthesisEventType.AUDIO,
+                    audio=tts.SynthesizedAudio(text="", data=audio_frame),
+                )
+            )
+
+    async def flush(self) -> None:
+        self._queue.put_nowait(self._text + " ")
+        self._text = ""
+        self._queue.put_nowait(STREAM_EOS)
+        await self._queue.join()
+
+    async def aclose(self) -> None:
+        self._main_task.cancel()
+        with contextlib.suppress(asyncio.CancelledError):
+            await self._main_task
+
+    async def __anext__(self) -> tts.SynthesisEvent:
+        if self._closed and self._event_queue.empty():
+            raise StopAsyncIteration
+
+        return await self._event_queue.get()
+
+
+def list_to_voices_list(data: list) -> List[Voice]:
+    voices = []
+    for voice in data:
+        voices.append(
+            Voice(
+                id=voice["id"],
+                name=voice["name"],
+                owner=voice["owner"],
+            )
+        )
+    return voices
+
+
+def _float_to_int16(array):
+    return np.clip(np.floor(array * 32767.5), -32768, 32767).astype(np.int16)
+
+
+def decode_mp3_to_frame(audio: bytes) -> rtc.AudioFrame:
+    tensor, sample_rate = torchaudio.load(io.BytesIO(audio), format="mp3")
```

Comment:
Do/Can we know ahead of time what the sample_rate will be?

When publishing an audio source in an agent, the sample rate (and num channels) must be known: https://github.com/livekit/agents/blob/main/examples/kitt/kitt.py#L76

It would be nice to publish at the same sample rate that lmnt is producing to avoid having to do a resample. This is especially impactful because, currently, LiveKit's resampler doesn't maintain continuity between subsequent audio frames so audio quality and artifacts can be introduced.

Reply:
Spoke about this internally - we may address this at the library level so we can leave this as-is